src/wp-includes/block-supports/elements.php

```diff
@@ -31,90 +31,79 @@ function wp_get_elements_class_name( $block ) {
-				'style.elements.h1.color.background',
-				'style.elements.h1.color.gradient',
-				'style.elements.h2.color.text',
-				'style.elements.h2.color.background',
-				'style.elements.h2.color.gradient',
-				'style.elements.h3.color.text',
-				'style.elements.h3.color.background',
-				'style.elements.h3.color.gradient',
-				'style.elements.h4.color.text',
-				'style.elements.h4.color.background',
-				'style.elements.h4.color.gradient',
-				'style.elements.h5.color.text',
-				'style.elements.h5.color.background',
-				'style.elements.h5.color.gradient',
-				'style.elements.h6.color.text',
-				'style.elements.h6.color.background',
-				'style.elements.h6.color.gradient',
-			),
-		),
-	);
+	$block_type       = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	$style_attributes = $block['attrs']['style']['elements'];
 
-	$skip_all_element_color_serialization = $element_color_properties['button']['skip'] &&
-		$element_color_properties['link']['skip'] &&
-		$element_color_properties['heading']['skip'];
+	// Button element support.
+	$supports_button  = ! wp_should_skip_block_supports_serialization( $block_type, 'color', 'button' );
+	$has_button_attrs = isset( $style_attributes['button']['color'] );
 
-	if ( $skip_all_element_color_serialization ) {
-		return $block_content;
-	}
+	if ( $supports_button && $has_button_attrs ) {
+		$button_attributes = $style_attributes['button']['color'];
 
-	$element_colors_set = 0;
+		if (
+			isset( $button_attributes['text'] ) ||
+			isset( $button_attributes['background'] ) ||
+			isset( $button_attributes['gradient'] )
+		) {
+			$tags = new WP_HTML_Tag_Processor( $block_content );
+			if ( $tags->next_tag() ) {
+				$tags->add_class( wp_get_elements_class_name( $block ) );
+			}
 
-	foreach ( $element_color_properties as $element_config ) {
-		if ( $element_config['skip'] ) {
-			continue;
+			return $tags->get_updated_html();
 		}
+	}
 
-		foreach ( $element_config['paths'] as $path ) {
-			if ( null !== _wp_array_get( $block['attrs'], explode( '.', $path ), null ) ) {
-				++$element_colors_set;
+	// Link element support.
+	$supports_link  = ! wp_should_skip_block_supports_serialization( $block_type, 'color', 'link' );
+	$has_link_attrs = isset( $style_attributes['link'] );
+
+	if ( $supports_link && $has_link_attrs ) {
+		if (
+			isset( $style_attributes['link']['color']['text'] ) ||
+			isset( $style_attributes['link'][':hover']['color']['text'] )
+		) {
+			$tags = new WP_HTML_Tag_Processor( $block_content );
+			if ( $tags->next_tag() ) {
+				$tags->add_class( wp_get_elements_class_name( $block ) );
 			}
+
+			return $tags->get_updated_html();
 		}
 	}
 
-	if ( ! $element_colors_set ) {
-		return $block_content;
-	}
+	// Heading element support.
+	$supports_heading = ! wp_should_skip_block_supports_serialization( $block_type, 'color', 'heading' );
+	if ( $supports_heading ) {
+		foreach ( $heading_elements as $element_name ) {
```

Comment:
```suggestion
                 $heading_elements = array( 'heading', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6' );
		foreach ( $heading_elements as $element_name ) {
```

Variable is only used here.